### PR TITLE
Backport 2.7: Fix compat.sh with ubuntu 16.04 gnutls 2.7

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Fix https://github.com/ARMmbed/mbedtls/issues/3520 (travis build) for 2.7. Backport of https://github.com/ARMmbed/mbedtls/pull/3590.

2.7 required more work due to a limitation in older versions of GnuTLS including our default one (3.4.10) but not the one we use as `GNUTLS_NEXT` (3.6.5): it compares distinguished names by their representation, so a `PrintableString` cannot be equal to a `UTF8String`. In 
https://github.com/ARMmbed/mbedtls/pull/1641 we patched some of our code and test data to cope with this limitation. In https://github.com/ARMmbed/mbedtls/pull/2418 we made a partial backport of the test data patches, but we didn't update the SHA-256 client certificate. This PR updates the whole `tests/data_files/cert_${hash}.crt` family to have `PrintableString` DN for the issuer.

The commit “cert_write: support all hash algorithms” is included here because it was necessary to regenerate the certificates. I made a separate PR for some other minor things that I stumbled on while working on this: https://github.com/ARMmbed/mbedtls/pull/3597.

### CODE REVIEW

**Review by mpg — APPROVED** [review]
Looks good to me. Thanks for investigating.

File: programs/x509/cert_write.c [hunk 1]
```diff
@@ -379,23 +380,14 @@ int main( int argc, char *argv[] )
...
-            else if( strcmp( q, "SHA512" ) == 0 )
-                opt.md = MBEDTLS_MD_SHA512;
-            else if( strcmp( q, "MD2" ) == 0 )
-                opt.md = MBEDTLS_MD_MD2;
-            else if( strcmp( q, "MD4" ) == 0 )
-                opt.md = MBEDTLS_MD_MD4;
-            else if( strcmp( q, "MD5" ) == 0 )
-                opt.md = MBEDTLS_MD_MD5;
-            else
+            const mbedtls_md_info_t *md_info =
```
Comment (hanno-arm) [inline programs/x509/cert_write.c:383]: :+1:

**Review by hanno-arm — CHANGES_REQUESTED** [review]
`library/certs.c` needs to be updated to contain the new `server2-sha256.crt` certificate. You might want to use #2596 for that. Other than that, the PR looks good to me.

**Comment by hanno-arm** [comment]
@gilles-peskine-arm @mpg https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.7/library/certs.c#L207 suggests that `library/certs.c` shouldn't be updated for ABI stability. What do you think?

**Review by mpg — APPROVED** [review]
LGTM

**Review by hanno-arm — APPROVED** [review]

**Comment by gilles-peskine-arm** [comment]
I rewrote the last commit because I'd accidentally committed a test version of `certs.c` with a changed size for `mbedtls_test_srv_crt_rsa`, instead of the version that I intended.

Regarding the conundrum of changing the certificate size vs keeping the file in `data_files` distinct from `certs.c`, this turned out not to be an issue because `certs.c` already had a different version as `data_files`, and the old version in `certs.c` had the same size as the new version in `data_files`. (Summarized from a chat discussion, and explained in the commit message.)

**Review by mpg — APPROVED** [review]
LGTM

**Review by hanno-arm — APPROVED** [review]

**Pull request merged**